### PR TITLE
Add CRUSH Rules API

### PR DESCRIPTION
**Pull request by shutstart**
- Adds the CRUSH Rules gRPC and REST endpoints to the ceph-api. 
- Basic tests for the listing, creation and deletion of crush rules.
- Removes hosts demo endpoint and its code.

### CODE REVIEW

File: api/crush_rule.proto [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    string name = 1;
+}
+
+// GET RULE
+message GetRuleRequest {
+    string name = 1;
+}
+
+message GetRuleResponse {
+    Rule rule = 1;
```
Comment (arttor) [inline api/crush_rule.proto:53]: Is it makes sense to just return `Rule` insteand of `GetRuleResponse`?

File: api/crush_rule.proto [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+message Step {
+    map<string, string> entries = 1; 
+}
+
+// CREATE RULE
+message CreateRuleRequest {
+    string device_class = 1;
+    string failure_domain = 2;
+    string name = 3;
```
Comment (arttor) [inline api/crush_rule.proto:35]: in [dashboard openapi](https://github.com/ceph/ceph/blob/7f9a2ec38f7d4f6d0c8d7698b0b8d4af45f5261b/src/pybind/mgr/dashboard/openapi.yaml#L4291-L4294) only name and failure_domain are required. Is it makes sense to make other fields `optional`.

File: api/crush_rule.proto [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    string pool_type = 4;
```
Comment (arttor) [inline api/crush_rule.proto:36]: consider using [enum](https://protobuf.dev/programming-guides/enum/) (replication, erasure) here. In this case it will be more clear which values are supported and what is default value.

File: pkg/api/crush_rule_api_handlers.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+	var cmdMap map[string]interface{}
+
+	if req.PoolType == "erasure" {
+		cmdMap = map[string]interface{}{
+			"prefix":  "osd crush rule create-erasure",
+			"name":    req.Name,
+			"profile": req.Profile,
+			"type":    req.FailureDomain,
+			"class":   req.DeviceClass,
```
Comment (arttor) [inline pkg/api/crush_rule_api_handlers.go:41]: optional parameters should be handled.

File: pkg/api/crush_rule_api_handlers.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+type crushRuleAPI struct {
+	radosSvc *rados.Svc
+}
+
+type crushDump struct {
+	Rules []*pb.Rule `json:"rules"`
+}
+
+// CreateRule implements pb.CrushRuleServer.
```
Comment (arttor) [inline pkg/api/crush_rule_api_handlers.go:27]: please remove autogenerated comments

File: pkg/api/crush_rule_api_handlers.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	// Find the rule by name.
+	for _, rule := range dump.Rules {
+		if rule.RuleName == req.Name {
+			return &pb.GetRuleResponse{Rule: rule}, nil
+		}
+	}
+
+	return nil, nil
```
Comment (arttor) [inline pkg/api/crush_rule_api_handlers.go:115]: can you please check how old api handles this case? should we return empty response or 404 error?

File: test/crush_rule_api_test.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+func Test_ListRlues(t *testing.T) {
+	r := require.New(t)
+	client := pb.NewCrushRuleClient(admConn)
+	res, err := client.ListRules(tstCtx, &emptypb.Empty{})
+	r.NoError(err)
+	rules := res.Rules
+	r.NotEmpty(rules)
+}
+
+func Test_Create_Get_And_Delete(t *testing.T) {
```
Comment (arttor) [inline test/crush_rule_api_test.go:20]: consider testing edge-cases:
- getting non-existing rule
- non-setting required fields
- ...

File: test/crush_rule_api_test.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	res, err := client.GetRule(tstCtx, &pb.GetRuleRequest{
+		Name: "some_rule",
+	})
+	r.NoError(err)
+	r.Equal("some_rule", res.Rule.RuleName)
+
+	_, err = client.DeleteRule(tstCtx, &pb.DeleteRuleRequest{
+		Name: "some_rule",
+	})
+	r.NoError(err)
```
Comment (arttor) [inline test/crush_rule_api_test.go:40]: please call `GetRule` once again to be sure that it was removed

File: test/crush_rule_api_test.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+		Root:          "default",
+		FailureDomain: "host",
+	})
+	r.NoError(err)
+
+	res, err := client.GetRule(tstCtx, &pb.GetRuleRequest{
+		Name: "some_rule",
+	})
+	r.NoError(err)
+	r.Equal("some_rule", res.Rule.RuleName)
```
Comment (arttor) [inline test/crush_rule_api_test.go:35]: please check that the rule appears in `ListRules` result.

**Review by arttor — CHANGES_REQUESTED** [review]
i've added some minor comments but overall looks pretty good!

File: api/crush_rule.proto (hunk 1 above)
Comment (shutstart) [reply api/crush_rule.proto:53]: Can return just the rule, make sense.

File: pkg/api/crush_rule_api_handlers.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+func (c *crushRuleAPI) CreateRule(ctx context.Context, req *pb.CreateRuleRequest) (*emptypb.Empty, error) {
+	if err := user.HasPermissions(ctx, user.ScopeOsd, user.PermCreate); err != nil {
+		return nil, err
+	}
+
+	// check if name and failure domain are set
+	if req.Name == "" || req.FailureDomain == "" {
+		return nil, types.ErrInvalidArg
+	}
```
Comment (arttor) [inline pkg/api/crush_rule_api_handlers.go:37]: Please add info for validation errors:
```suggestion
	if req.Name == "" {
		return nil, fmt.Errorf("%w: name is required", types.ErrInvalidArg)
	}
	if req.FailureDomain == "" {
		return nil, fmt.Errorf("%w: failure domain is required", types.ErrInvalidArg)
	}
```

**Review by arttor — CHANGES_REQUESTED** [review]

**Review by arttor — APPROVED** [review]
